### ISSUE TITLE
Re-enable --enable-hardening security features

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#915]
When we added the `-O0` flag to disable optimization, that disabled `FORTIFY_SOURCE` security features. You can see the warnings saying it requires optimization to be enabled in the build output. We have to re-enable optimization (and make sure `--enable-hardening` is on) before we release 1.0.

**Comment by nathan-at-least** [comment]
I put this in the Consensus Code Security Freeze milestone and removed the engineering agenda label. Feel free to re-add the label if I missed something.

**Comment by defuse** [comment]
I think I may have fixed this when I fixed the build warnings.

**Comment by defuse** [comment]
Let's close this only after I've found a way to verify all the security features we want turned on, and that's been automated and put into the test suite.

**Comment by defuse** [comment]
The PR for this is #1064.

**Comment by defuse** [comment]
Here's a post listing the available security features: https://security.stackexchange.com/questions/24444/what-is-the-most-hardened-set-of-options-for-gcc-compiling-c-c

Note that FORTIFY_SOURCE is only fully enabled at the `-O2` level (and above?).

**Comment by defuse** [comment]
We want these on our *dependency* builds too, not just the main Zcash build.

**Comment by defuse** [comment]
In scope for this ticket:

- Stack protector (canaries), strong or all.
- ASLR (make sure everything, even `.text` is randomized).
- FORTIFY_SOURCE is at level 2.
- Full relro (if possible, if not then partial).
- Try -Wconversion, -Wsign-conversion on the main build, and see if it works.

Next step, figure out how feasible it is to look at the binary and check which of them are turned on.

**Comment by defuse** [comment]
ASLR test can be done by loading it twice and checking that the virtual address changes. RELRO can also be tested by loading and examining the memory segments of the running process.

Stack protector and FORTIFY_SOURCE probably can't be tested without writing a script that looks at the assembly, which is probably too much of a maintenance burden.

The flags are obviously compile-time things and can't be tested by looking at the binary.

**Comment by defuse** [comment]
Relevant: https://blogs.cisco.com/security/how_was_this_executable_built

**Comment by defuse** [comment]
The checksec.sh script from that blog post is awesome!

```
[firexware@firexware-pc zcash]$ ./checksec.sh --file src/zcashd
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      FILE
Full RELRO      Canary found      NX enabled    PIE enabled     No RPATH   No RUNPATH   src/zcashd
```

```
[firexware@firexware-pc zcash]$ ./checksec.sh --file /bin/ls
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      FILE
Partial RELRO   Canary found      NX enabled    No PIE          No RPATH   No RUNPATH   /bin/ls
```

**Comment by defuse** [comment]
It can also check for FORTIFY_SOURCE:

```
$ ./checksec.sh --fortify-file src/zcashd
* FORTIFY_SOURCE support available (libc)    : Yes
* Binary compiled with FORTIFY_SOURCE support: Yes

 ------ EXECUTABLE-FILE ------- . -------- LIBC --------
 FORTIFY-able library functions | Checked function names
 -------------------------------------------------------
 printf_chk                     | __printf_chk
 printf                         | __printf_chk
 printf_chk                     | __printf_chk   <-- each one is either red or green
 printf                         | __printf_chk
 readlink                       | __readlink_chk
 readlink                       | __readlink_chk
...
```

It's BSD licensed so I'll see if I can make it produce machine-readable output (instead of colored text) and then add it to our test suite.

**Comment by daira** [comment]
I'm concerned that the use of [AX_CHECK_PREPROC_FLAG](https://www.gnu.org/software/autoconf-archive/ax_check_preproc_flag.html) in https://github.com/zcash/zcash/blob/zc.v0.11.2.latest/configure.ac#L411 means that ``-Wstack-protector`` and ``-D_FORTIFY_SOURCE=2`` will silently not be used if their use results in a compiler error.

**Comment by defuse** [comment]
We've determined that `checksec.sh` is not sufficient to check that FORTIFY_SOURCE is turned on for the entire binary. Here's what it looks like when we pass `--disable-hardening`:

```
$ ./qa/zcash/check-security-hardening.sh 
FAIL: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcashd is missing basic hardening features.
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      FILE
No RELRO        Canary found      NX enabled    No PIE          No RPATH   No RUNPATH   /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcashd
FAIL: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcash-cli is missing basic hardening features.
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      FILE
No RELRO        No canary found   NX enabled    No PIE          No RPATH   No RUNPATH   /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcash-cli
FAIL: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcash-gtest is missing basic hardening features.
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      FILE
No RELRO        Canary found      NX enabled    No PIE          No RPATH   No RUNPATH   /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcash-gtest
FAIL: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/bitcoin-tx is missing basic hardening features.
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      FILE
No RELRO        Canary found      NX enabled    No PIE          No RPATH   No RUNPATH   /home/firexware/Data-1/Documents/FireXware/la/zcash/src/bitcoin-tx
PASS: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcashd has FORTIFY_SOURCE.
FAIL: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcash-cli is missing FORTIFY_SOURCE.
PASS: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/zcash-gtest has FORTIFY_SOURCE.
PASS: /home/firexware/Data-1/Documents/FireXware/la/zcash/src/bitcoin-tx has FORTIFY_SOURCE.
```

My best guess at explaining this is that `checksec.sh` works by looking for symbols related to FORTIFY_SOURCE, and that one of the dependencies of `zcashd` uses its own FORTIFY_SOURCE which creates a false positive pass, and `zcash-cli` does not include that dependency.

**Comment by defuse** [comment]
(Or, at least the simple grep-the-output test we're doing isn't sufficient; checksec.sh produces more output which might help us tell if it's on for the whole binary or not.)

**Comment by daira** [comment]
Can you paste the full ``checksec.sh`` output for the ``--disable-hardening`` case please?

**Comment by defuse** [comment]
With `--disable-hardening` (all are red except for the two `memcpy_chk` lines which are green).

```
[firexware@firexware-pc zcash]$ ./qa/zcash/checksec.sh --fortify-file src/zcashd | tee disable-hardening-output.txt 
* FORTIFY_SOURCE support available (libc)    : Yes
* Binary compiled with FORTIFY_SOURCE support: Yes

 ------ EXECUTABLE-FILE ------- . -------- LIBC --------
 FORTIFY-able library functions | Checked function names
 -------------------------------------------------------
 printf                         | __printf_chk
 printf                         | __printf_chk
 readlink                       | __readlink_chk
 readlink                       | __readlink_chk
 read                           | __read_chk
 read                           | __read_chk
 fread_unlocked                 | __fread_unlocked_chk
 fread_unlocked                 | __fread_unlocked_chk
 pread                          | __pread_chk
 pread                          | __pread_chk
 getcwd                         | __getcwd_chk
 getcwd                         | __getcwd_chk
 fprintf                        | __fprintf_chk
 fprintf                        | __fprintf_chk
 wmemcpy                        | __wmemcpy_chk
 wmemcpy                        | __wmemcpy_chk
 memmove                        | __memmove_chk
 memmove                        | __memmove_chk
 vsnprintf                      | __vsnprintf_chk
 vsnprintf                      | __vsnprintf_chk
 fgets                          | __fgets_chk
 fgets                          | __fgets_chk
 strncpy                        | __strncpy_chk
 strncpy                        | __strncpy_chk
 snprintf                       | __snprintf_chk
 snprintf                       | __snprintf_chk
 memset                         | __memset_chk
 memset                         | __memset_chk
 recv                           | __recv_chk
 recv                           | __recv_chk
 memcpy                         | __memcpy_chk
 memcpy_chk                     | __memcpy_chk
 memcpy                         | __memcpy_chk
 memcpy_chk                     | __memcpy_chk
 fread                          | __fread_chk
 fread                          | __fread_chk
 poll                           | __poll_chk
 poll                           | __poll_chk
 vfprintf                       | __vfprintf_chk
 vfprintf                       | __vfprintf_chk
 strcpy                         | __strcpy_chk
 strcpy                         | __strcpy_chk
 sprintf                        | __sprintf_chk
 sprintf                        | __sprintf_chk
 strcat                         | __strcat_chk
 strcat                         | __strcat_chk
 recvfrom                       | __recvfrom_chk
 recvfrom                       | __recvfrom_chk

SUMMARY:

* Number of checked functions in libc                : 78
* Total number of library functions in the executable: 21711
* Number of FORTIFY-able functions in the executable : 48
* Number of checked functions in the executable      : 2
* Number of unchecked functions in the executable    : 46
```

**Comment by defuse** [comment]
With `--enable-hardening` there are significantly more green ones (printf_chk x 2, fdelt_chk x 2, fprintf_chk x 2, vsnprintf_chk x 2, snprintf_chk x 2, memcpy_chk x 2, sprintf_chk x 2, strcat_chk x 2):

```
[firexware@firexware-pc zcash]$ ./qa/zcash/checksec.sh --fortify-file src/zcashd | tee enable-hardening-output.txt 
* FORTIFY_SOURCE support available (libc)    : Yes
* Binary compiled with FORTIFY_SOURCE support: Yes

 ------ EXECUTABLE-FILE ------- . -------- LIBC --------
 FORTIFY-able library functions | Checked function names
 -------------------------------------------------------
 printf_chk                     | __printf_chk
 printf                         | __printf_chk
 printf_chk                     | __printf_chk
 printf                         | __printf_chk
 readlink                       | __readlink_chk
 readlink                       | __readlink_chk
 fdelt_chk                      | __fdelt_chk
 fdelt_chk                      | __fdelt_chk
 read                           | __read_chk
 read                           | __read_chk
 fread_unlocked                 | __fread_unlocked_chk
 fread_unlocked                 | __fread_unlocked_chk
 pread                          | __pread_chk
 pread                          | __pread_chk
 getcwd                         | __getcwd_chk
 getcwd                         | __getcwd_chk
 fprintf                        | __fprintf_chk
 fprintf_chk                    | __fprintf_chk
 fprintf                        | __fprintf_chk
 fprintf_chk                    | __fprintf_chk
 wmemcpy                        | __wmemcpy_chk
 wmemcpy                        | __wmemcpy_chk
 memmove                        | __memmove_chk
 memmove                        | __memmove_chk
 vsnprintf                      | __vsnprintf_chk
 vsnprintf_chk                  | __vsnprintf_chk
 vsnprintf                      | __vsnprintf_chk
 vsnprintf_chk                  | __vsnprintf_chk
 fgets                          | __fgets_chk
 fgets                          | __fgets_chk
 strncpy                        | __strncpy_chk
 strncpy                        | __strncpy_chk
 snprintf_chk                   | __snprintf_chk
 snprintf                       | __snprintf_chk
 snprintf_chk                   | __snprintf_chk
 snprintf                       | __snprintf_chk
 memset                         | __memset_chk
 memset                         | __memset_chk
 recv                           | __recv_chk
 recv                           | __recv_chk
 memcpy                         | __memcpy_chk
 memcpy_chk                     | __memcpy_chk
 memcpy                         | __memcpy_chk
 memcpy_chk                     | __memcpy_chk
 fread                          | __fread_chk
 fread                          | __fread_chk
 poll                           | __poll_chk
 poll                           | __poll_chk
 vfprintf                       | __vfprintf_chk
 vfprintf                       | __vfprintf_chk
 strcpy                         | __strcpy_chk
 strcpy                         | __strcpy_chk
 sprintf                        | __sprintf_chk
 sprintf_chk                    | __sprintf_chk
 sprintf                        | __sprintf_chk
 sprintf_chk                    | __sprintf_chk
 strcat_chk                     | __strcat_chk
 strcat                         | __strcat_chk
 strcat_chk                     | __strcat_chk
 strcat                         | __strcat_chk
 recvfrom                       | __recvfrom_chk
 recvfrom                       | __recvfrom_chk

SUMMARY:

* Number of checked functions in libc                : 78
* Total number of library functions in the executable: 21760
* Number of FORTIFY-able functions in the executable : 62
* Number of checked functions in the executable      : 16
* Number of unchecked functions in the executable    : 46
```

**Comment by defuse** [comment]
I wish I understood FORTIFY_SOURCE enough to know what the above means... maybe we can do things to get even more protection?

Shall I test that "Number of checked functions in the executable" is 16?

**Comment by daira** [comment]
Maybe check that it's at least 16. That will at least catch regressions (unless the number increases first).

**Comment by defuse** [comment]
Let's contribute these tests to upstream!

**Comment by daira** [comment]
Can I just have a bit of a rant: being forced to trade off between the FORTIFY_SOURCE protections and a greater probability of compiler optimization bugs makes absolutely no sense. There is no good reason why FORTIFY_SOURCE should have been dependent on ``-O2``; it's just lazy implementation.

**Comment by daira** [comment]
Actually are we sure is is dependent on ``-O2``? I see some places that only say ``-O1``, e.g. http://man7.org/linux/man-pages/man7/feature_test_macros.7.html :

> If ``_FORTIFY_SOURCE`` is set to 1, with compiler optimization level 1 (``gcc -O1``) and above, checks that shouldn't change the behavior of conforming programs are performed.  With ``_FORTIFY_SOURCE`` set to 2, some more checking is added, but some conforming programs might fail.

This doesn't imply that ``-O2`` is needed for ``_FORTIFY_SOURCE=2``; as I read it, only ``-O1`` or higher is needed for any ``-FORTIFY_SOURCE`` > 0. And some of the compiler optimizations added at ``-O2`` are quite scarily complicated-sounding: https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

**Comment by zookozcash** [comment]
https://github.com/zcash/zcash/issues/1071

**Comment by defuse** [comment]
I think you're right, we only need `-O1`. Debian docs https://wiki.debian.org/Hardening#DEB_BUILD_HARDENING_FORTIFY_.28gcc.2Fg.2B-.2B-_-D_FORTIFY_SOURCE.3D2.29 say "Note that for this feature to be fully enabled, the source must also be compiled with -O1 or higher." (but their examples use -O2). I thought we needed O2 because some random person on stackoverflow said "I have also read that -D_FORTIFY_SOURCE=2 should be used with -O2, otherwise not all features would be available." They didn't give a reference, and I can't find one.

**Comment by ebfull** [comment]
Nota bene libzcash CPP flags.

**Comment by defuse** [comment]
There's a thing in upstream's latest version called `make check-security`, which does some very similar things. It even has tests written for it (tests that test the test) so let's see how hard it will be to bring it in here and add the missing stuff.

**Comment by daira** [comment]
+1 for integrating this into ``make check-security``, and let's use ``-O1`` (or at least check how the benchmarks look with ``-O1``).

**Comment by defuse** [comment]
Here's the list of flags and things that Arch Linux is considering adopting: https://github.com/pid1/test-sec-flags/blob/master/FLAGS.rst

**Comment by defuse** [comment]
Those AC_CHECK_LINK_FLAG was hiding at least one failure: `-Wl,--dynamicbase`. I think that's a windows-only thing? It doesn't seem to be available in my linker.

**Comment by defuse** [comment]
```
--dynamicbase
    The image base address may be relocated using address space layout randomization ( ASLR ). This feature was introduced with MS Windows Vista for i386 PE targets. 
```

http://linux.die.net/man/1/ld

So I guess I can move that into the windows-only options.

**Comment by defuse** [comment]
Same thing for

```
--nxcompat
    The image is compatible with the Data Execution Prevention. This feature was introduced with MS Windows XP SP2 for i386 PE targets. 
```

**Comment by defuse** [comment]
I can't add `-Wconversion -Wsign-conversion` on the main build because it causes compilation of the boost test program hard-coded into `configure.ac` to fail. Maybe I can fix `configure.ac` so that it uses a default `CXXFLAGS` to try that test program.

**Comment by defuse** [comment]
```
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/date_time/wrapping_int.hpp:151:14: error: conversion to 'boost::date_time::wrapping_int2<short int, 1, 12>::int_type {aka short int}' from 'int' may alter its value [-Werror=conversion]
       value_ -= (wrap_max - wrap_min + 1);
       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/date_time/wrapping_int.hpp:156:14: error: conversion to 'boost::date_time::wrapping_int2<short int, 1, 12>::int_type {aka short int}' from 'int' may alter its value [-Werror=conversion]
       value_ += (wrap_max - wrap_min + 1);
       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
configure:26641: $? = 1
configure: failed program was:
```

**Comment by defuse** [comment]
Okay nope, there are 100s of errors with `-Wconversion -Wsign-conversion`. I'll try enabling them for just libzcash.

**Comment by defuse** [comment]
That doesn't work either, there are too many (dangerous-looking) warnings coming from libsnark headers, e.g.

```
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/libsnark/algebra/fields/fp.tcc:175:56: error: conversion to ‘mp_size_t {aka long int}’ from ‘long unsigned int’ may change the sign of the result [-Werror=sign-conversion]
             carryout = mpn_add_1(res+n+i, res+n+i, n-i, carryout);
                                                       ~^~
In file included from /home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/format/alt_sstream.hpp:173:0,
                 from /home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/format/internals.hpp:23,
                 from /home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/format.hpp:38,
                 from zcash/JoinSplit.cpp:10:
```

So I opened #1266.

**Comment by defuse** [comment]
Everything in the scope of this ticket (https://github.com/zcash/zcash/issues/915#issuecomment-230139318) is done, has an open PR, or has been moved into its own ticket.
